commercial/app/model/commercial/CapiAgent.scala

```diff
@@ -0,0 +1,47 @@
+package model.commercial
+
+import common.AkkaAgent
+import model.Content
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.control.NonFatal
+
+object CapiAgent {
+
+  private lazy val shortUrlAgent = AkkaAgent[Map[String, Option[Content]]](Map.empty)
+
+  def contentByShortUrls(shortUrls: Seq[String])
+                        (implicit ec: ExecutionContext): Future[Seq[Content]] = {
+
+    val shortUrlIds = shortUrls map { url =>
+      val slashPrefixed = if (url startsWith "/") url else s"/$url"
+      val endTrimmed = slashPrefixed.take(slashPrefixed.indexOf("/", 3))
+      endTrimmed
+    }
+
+    val cache = shortUrlAgent.get()
+
+    val urlsNotInCache = shortUrlIds filterNot cache.contains
+
+    def addToCache(contents: Seq[Content]): Future[Map[String, Option[Content]]] = {
+      val initialValues = urlsNotInCache.map(_ -> None).toMap
+      shortUrlAgent alter (_ ++ initialValues)
+      val newContents = contents.map(content => content.shortUrlId -> Some(content)).toMap
+      shortUrlAgent alter (_ ++ newContents)
+    }
+
+    val eventualNewCache = if (urlsNotInCache.isEmpty) {
+      Future.successful(cache)
+    } else {
+      Lookup.contentByShortUrls(shortUrlIds) flatMap {
+        addToCache
+      } recoverWith {
+        case NonFatal(e) => Future.successful(cache)
```

Comment:
I take your point.  I should be logging that exception.  There ought to be a better highlight in the UI of which line you're commenting on.